documentation/src/main/asciidoc/manual-index.asciidoc

```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
+        .restrictedBy( Restrictions.in( "name", "Foo", "Bar" ) )
+        .build();
+----
+====
+
+==== Parallelism
+In order to maximize the performance, we highly recommend you to speed up the mass indexer using
+parallelism. Parallelism is activated by default. Under the standard JSR 352, the exact word is
+_"partitioning"_: The indexing step may run as multiple partitions, one per thread. Each partition
+has its own partition ID and parameters. If there’re more partitions than threads, partitions are\
+considered as a queue to consume: each thread can only run one partition at a time and won’t consume
+the next partition until the previous one is finished.
+
+===== Threads
+The maximum number of threads used by the job execution is defined through method `maxThreads()`.
+Within the N threads given, there’s 1 thread reserved for the core, so only N - 1 threads are
+available for different partitions. The default number of threads used in Hibernate Search is 10,
+which correspond to the max number of threads in JBeret. You can overwrite you maximum number used
+by Hibernate Search batch job, but we cannot guarantee that your batch runtime can provide the
+requested number.
+
+====
+[source, JAVA]
+----
+MassIndexingJob.parameters()
+        .maxThreads( 5 )
+        ...
+----
+====
+
+[NOTE]
+====
+Note that the batch runtime cannot guarantee the requested number of threads are available, it will
+use as it can up to the requested maximum. (JSR352 v1.0 Final Release, page 34)
+====
+
+This approach is optimized for best performance but requires to set the application in maintenance
+mode. Running a mass indexer in busy hours is not recommended. You also need to ensure the batch
+runtime has enough available threads for the execution, because all the batch jobs share the same
+thread pool.
+
+===== Row per partitions
+Mass indexer defines dynamically the number of partitions, but you need to tell it the number of
+rows per partitions so that it can calculate the partitions by division. Each partition deals with
+one root entity type, so 2 different entity types will never run under the same partition.
+
+====
+[source, JAVA]
+----
+MassIndexingJob.parameters()
+        .rowsPerPartition( 5000 )
+        ...
+----
+====
+
+==== Checkpoint
+Mass indexer supports checkpoint algorithm, which helps you to restart your suspended or failed job
+from the last successful checkpoint. Assume that N is the value of checkpoint interval, then a
+partition will reach the checkpoint at every N items processed. You can overwrite the default
+interval to adapt your business requirement. By default, it is set to 500.
+
+====
+[source, JAVA]
+----
+MassIndexingJob.parameters()
+        .checkpointInterval( 1000 )
+        ...
+----
+====
+
+Checkpoint a partition-scope algorithm, which means each partition has its own checkpoint counter
+and it is not shared with other partitions.
+
+
+==== Selecting the persistence unit (EntityManagerFactory)
```

Comment:
I should restructure this section. Don't do it, I will, but I'm adding this comment as a reminder.

In particular:

 * the "Plain Java environment" section should go last
 * I should add a section about "other dependency injection mechanisms"
 * I should explain what I mean by "Plain Java environment" more precisely. In particular, it's not about "no CDI" but "no dependency injection mechanism".